kernel/src/sched.rs

```diff
@@ -364,7 +374,9 @@ impl Kernel {
                             process.set_fault_state();
                         }
                         Some(ContextSwitchReason::SyscallFired { syscall }) => {
-                            process.debug_syscall_called(syscall);
+                            if config::CONFIG.debug_processes {
+                                process.debug_syscall_called(syscall);
+                            }
```

Comment:
If so, it might be more straightforward to relocate all of the `if config...` checks into the debug functions themselves rather than having them sprinkled throughout the code.

Reply:
I actually tried this before creating this PR but tried again after implementing @bradjc's comment below to make sure. Benchmarking something like this is tricky but IIUC it doesn't work for functions that are on the trait but it works when they are implemented on the struct. 
This conclusion is not based on the number of cycles, because it depends on compiler settings and can be misleading, but `objdump` output. I see calls to `debug_*` functions that are on the trait when `debug_processes = false` (even when they have an early return) unless they are wrapped in an if statement. Thus, I was able to move the if statements into the `debug_*` functions implemented on the struct but kept the rest (for calls outside process.rs). I hope this is an acceptable compromise.

Reply:
I don't think functions on the `ProcessType` trait can be effectively optimized away by the compiler, because this trait is currently used as trait objects, that have dynamic dispatch (even though in practice there's only one type per compiled board).

If you're interested, I've started documenting this overhead in #1761 (see also #1663).
Removing `&dyn` on some GPIOs types recently led to code size improvements in #1765.

Reply:
I find it natural that the functions on the trait remain, but I am not sure why the calls are not optimized away. Maybe they are optimized in other cases as a side-effect? Thanks for the issue numbers, I'll take a look!